distribution/lib/Standard/Base/0.0.0-dev/src/Data/Ordering/Natural_Order.enso

```diff
@@ -0,0 +1,65 @@
+from Standard.Base import all
+
+import Standard.Base.Data.Text.Regex
+import Standard.Base.Data.Text.Regex.Mode
+import Standard.Base.Data.Ordering.Vector_Lexicographic_Order
+
+## Compares two text values according to the natural dictionary ordering.
+
+   > Example
+     Compare two texts according to the natural dictionary ordering.
+
+         Natural_Order.compare "a2" "a100" == Ordering.Less
+
+   > Example
+     Sort a vector of texts according to the natural dictionary ordering.
+
+         ["a2", "a1", "a100", "a001", "a0001"].sort by=Natural_Order.compare . should_equal ["a0001", "a001", "a1", "a2", "a100"]
+compare : Text -> Text -> Ordering
+compare text1 text2 =
+    nat1 = here.to_natural_key text1
+    nat2 = here.to_natural_key text2
+    ## TODO [RW] The additional assignment to a temporary variable is a
+       workaround for the following bug:
+       https://www.pivotaltracker.com/story/show/181162108
+    res = Vector_Lexicographic_Order.compare nat1 nat2
+    res
+
+## PRIVATE
+to_natural_key : Text -> Vector
+to_natural_key text = if text.is_empty then [] else
+    splitter_regex = Regex.compile "([^0-9]+|[0-9]+)"
+    parts = splitter_regex.find text mode=Mode.All
+
+    is_integer_regex = Regex.compile "[0-9]+"
+    parts.map part-> case is_integer_regex.matches part of
+        True ->
+            value = Decimal.parse part . floor
+            Natural_Ordering_Numeric_Part value part
+        False ->
+            Natural_Ordering_Alphabetical_Part part
+
+## PRIVATE
+type Natural_Ordering_Part
+    ## PRIVATE
+    type Natural_Ordering_Alphabetical_Part (text : Text)
+
+    ## PRIVATE
+    type Natural_Ordering_Numeric_Part (value : Integer) (original_text : Text)
+
+    ## PRIVATE
+    compare_to : Natural_Ordering_Part -> Ordering
+    compare_to that = case Pair this that of
```

Comment:
It is valid to compare `abc` to 123.
Lets for the sake of convention say that a number < a string
(do a quick check on the definitions to see if should be >?)